Harden git_sitrep()

Test it under more edge cases.

Here's one seen in CI:

```
> git_sitrep()
Git config (global)
* Name: 'GitHub Actions User'
* Email: 'ghau@example.com'
* Vaccinated: FALSE
usethis + gert
* Default usethis protocol: <unset>
* gert supports HTTPS: TRUE
* gert supports SSH: TRUE
GitHub
* Personal access token: '<found in env var>'
✖ Can't validate token. Is the network reachable?
✖ Can't validate token. Is the network reachable?
Repo
* Path: '/Users/runner/runners/2.263.0/work/usethis/usethis'
* Local branch -> remote tracking branch: '<unset>' -> '<unset>'
GitHub pull request readiness
##[error]Error: Failed to validate the GitHub `auth_token`
GitHub API error (403): 403 Forbidden
Message: Resource not accessible by integration
Read more at https://developer.github.com/v3/users/#get-the-authenticated-user
```

--- a/R/git.R
+++ b/R/git.R
@@ -478,19 +478,14 @@ git_sitrep <- function() {
   # GitHub remote config -------------------------------------------------------
   cfg <- github_remote_config()
   repo_host <- cfg$host_url
-  if (repo_host != default_gh_host) {
+  if (!is.na(repo_host) && repo_host != default_gh_host) {
     kv_line("Non-default GitHub host", repo_host)
     pat_found <- pat_sitrep(host = repo_host)
   }
 
   hd_line("GitHub remote configuration")
-  if (cfg$type == "no_github") {
-    ui_info("
-      This repo has neither {ui_value('origin')} nor {ui_value('upstream')} \\
-      remote on GitHub")
-    return(invisible())
-  }
-  print(cfg)
+  purrr::walk(format(cfg), ~ ui_bullet(.x, cli::symbol$star))
+  invisible()
 }
 
 pat_sitrep <- function(host = "https://github.com") {

--- a/R/git.R
+++ b/R/git.R
@@ -481,8 +481,11 @@ git_sitrep <- function() {
           )
         }
       },
-      http_error_404 = function(e) kv_line("Email(s)", "<unknown>"),
-      error = function(e) ui_oops("Can't retrieve user's GitHub email addresses")
+      error = function(e) {
+        ui_oops("
+          Can't retrieve registered email addresses
+          If you are troubleshooting, check GitHub host, token, and token scopes")
+      }
     )
   } else {
     kv_line("Personal access token", NULL)
@@ -490,7 +493,20 @@ git_sitrep <- function() {
       Call {ui_code('gh_token_help()')} for help configuring a token")
   }
 
-  # git project ---------------------------------------------------------------
+  # git and github for active project ------------------------------------------
+  hd_line("Git repo associated with the current project")
+  ui_silence(try(proj_get(), silent = TRUE))
+  if (!proj_active()) {
+    ui_info("No active usethis project")
+    return(invisible())
+  }
+  kv_line("Active usethis project", proj_get())
+  if (!uses_git()) {
+    ui_info("Active project is not a Git repo")
+    return(invisible())
+  }
+
+  # local git config -----------------------------------------------------------
   if (proj_active() && uses_git()) {
     local_user <- list(
       user.name = git_cfg_get("user.name", "local"),
@@ -503,32 +519,19 @@ git_sitrep <- function() {
     }
   }
 
-  # repo overview -------------------------------------------------------------
-  hd_line("Repo")
-  ui_silence(try(proj_get(), silent = TRUE))
-  if (!proj_active()) {
-    ui_info("No active usethis project.")
-    return(invisible())
-  }
-
-  if (!uses_git()) {
-    ui_info("Active project is not a Git repo.")
-    return(invisible())
-  }
-
-  kv_line("Path", git_repo())
+  # current branch -------------------------------------------------------------
   branch <- tryCatch(git_branch(), error = function(e) NULL)
   tracking_branch <- if (is.null(branch)) NA_character_ else git_branch_tracking()
-  ## TODO: rework when ui_*() functions make it possible to do better
+  # TODO: can't really do what I want with existing ui_*() functions
+  # e.g., vertical alignment would make this nicer
   branch <- if (is.null(branch)) "<unset>" else branch
   tracking_branch <- if (is.na(tracking_branch)) "<unset>" else tracking_branch
-  ui_inform(
-    "* ", "Local branch -> remote tracking branch: ",
-    ui_value(branch), " -> ", ui_value(tracking_branch)
-  )
+  ui_inform(glue("
+    * Local branch -> remote tracking branch:
+      {ui_value(branch)} -> {ui_value(tracking_branch)}"))
 
   # PR outlook -------------------------------------------------------------
-  hd_line("GitHub pull request readiness")
+  hd_line("GitHub remote configuration")
   # TODO: need to surface host more here
   cfg <- github_remote_config()
   if (cfg$type == "no_github") {

--- a/R/git.R
+++ b/R/git.R
@@ -438,14 +438,6 @@ git_sitrep <- function() {
     }
   }
 
-  # usethis + gert + credentials -----------------------------------------------
-  hd_line("usethis + gert")
-  kv_line("Default usethis protocol", getOption("usethis.protocol"))
-  kv_line("gert supports HTTPS", gert::libgit2_config()$https)
-  kv_line("gert supports SSH", gert::libgit2_config()$ssh)
-  # TODO: forward more info from the credentials package when available
-  # https://github.com/r-lib/credentials/issues/6
-
   # github ---- ----------------------------------------------------------------
   hd_line("GitHub")
   default_gh_host <- get_hosturl(default_api_url())

--- a/R/git.R
+++ b/R/git.R
@@ -441,20 +441,22 @@ git_sitrep <- function() {
         kv_line("GitHub user", who$login)
         scopes <- who$scopes
         kv_line("Token scopes", who$scopes)
+        # https://docs.github.com/en/free-pro-team@latest/developers/apps/scopes-for-oauth-apps
         # why these checks?
         # previous defaults for create_github_token(): repo, gist, user:email
         # more recently: repo, user, gist
         # (gist scope is a very weak recommendation)
         scopes <- strsplit(scopes, ", ")[[1]]
-        if (!any(grepl("^repo$", scopes)) ||
+        if (length(scopes) == 0 ||
+            !any(grepl("^repo$", scopes)) ||
             !any(grepl("^user(:email)?$", scopes))) {
           ui_oops("
-            Token may be mis-scoped? {ui_value('repo')} and \\
+            Token may be mis-scoped: {ui_value('repo')} and \\
             {ui_value('user')} are highly recommended scopes
             If you are troubleshooting, consider this")
         }
       },
-      http_error_401 = function(e) ui_oops("Token is invalid."),
+      http_error_401 = function(e) ui_oops("Token is invalid"),
       error = function(e) {
         ui_oops("
           Can't get user profile for this token. Is the network reachable?")

--- a/R/git.R
+++ b/R/git.R
@@ -424,6 +424,8 @@ git_sitrep <- function() {
   if (!vaccinated) {
     ui_info("See {ui_code('?git_vaccinate')} to learn more about this")
   }
+  # TODO: Revisit when I harden the HTTPS default
+  kv_line("Default usethis protocol", getOption("usethis.protocol"))
 
   # git project ---------------------------------------------------------------
   if (proj_active() && uses_git()) {

--- a/R/git.R
+++ b/R/git.R
@@ -435,12 +435,12 @@ git_sitrep <- function() {
     )
     if (!is.null(local_user$user.name) || !is.null(local_user$user.name)) {
       hd_line("Git config (project)")
-      kv_line("Name", git_cfg_get("user.name"))
-      kv_line("Email", git_cfg_get("user.email"))
+      kv_line("Name", local_user$user.name)
+      kv_line("Email", local_user$user.email)
     }
   }
 
-  # github ---- ----------------------------------------------------------------
+  # github ---------------------------------------------------------------------
   hd_line("GitHub")
   default_gh_host <- get_hosturl(default_api_url())
   kv_line("Default GitHub host", default_gh_host)

--- a/R/git.R
+++ b/R/git.R
@@ -442,32 +442,62 @@ git_sitrep <- function() {
   # TODO: forward more info from the credentials package when available
   # https://github.com/r-lib/credentials/issues/6
 
-  # github user ---------------------------------------------------------------
+  # github ---- ----------------------------------------------------------------
   hd_line("GitHub")
-  auth_token <- gh::gh_token()
+  default_gh_host <- get_hosturl(default_api_url())
+  kv_line("Default GitHub host", default_gh_host)
+  auth_token <- gh::gh_token(api_url = default_gh_host)
   have_token <- auth_token != ""
   if (have_token) {
     kv_line("Personal access token", "<discovered>")
     tryCatch(
       {
-        who <- gh::gh_whoami(auth_token)
+        who <- gh::gh_whoami(.token = auth_token, .api_url = default_gh_host)
         kv_line("User", who$login)
         kv_line("Name", who$name)
+        scopes <- who$scopes
+        kv_line("Scopes", who$scopes)
+        # previous defaults for create_github_token(): repo, gist, user:email
+        # more recently: repo, user, gist
+        # and gist scope is a very weak recommendation
+        # hence the checks aren't exactly for our defaults
+        scopes <- strsplit(scopes, ", ")[[1]]
+        if (!any(grepl("^repo$", scopes)) ||
+            !any(grepl("user(:email)?", scopes))) {
+          ui_oops("
+            Token may be underscoped: {ui_value('repo')} and \\
+            {ui_value('user')} are highly recommended scopes")
+        }
       },
       http_error_401 = function(e) ui_oops("Token is invalid."),
       error = function(e) ui_oops("Can't validate token. Is the network reachable?")
     )
     tryCatch(
       {
-        emails <- unlist(gh::gh("/user/emails", .token = auth_token))
-        emails <- emails[names(emails) == "email"]
-        kv_line("Email(s)", emails)
+        emails <- gh::gh(
+          "/user/emails",
+          .token = auth_token,
+          .api_url = default_gh_host
+        )
+        addresses <- map_chr(
+          emails,
+          ~ if (.x$primary) glue_data(.x, "{email} (primary)") else .x[["email"]]
+        )
+        kv_line("Email(s)", addresses)
+        global_email <- git_cfg_get("user.email", "global")
+        if (!is.null(global_email) && !grepl(global_email, addresses)) {
+          ui_oops(
+            "User's local Git email doesn't appear to be registered with GitHub"
+          )
+        }
       },
       http_error_404 = function(e) kv_line("Email(s)", "<unknown>"),
-      error = function(e) ui_oops("Can't validate token. Is the network reachable?")
+      error = function(e) ui_oops("Can't retrieve user's GitHub email addresses")
     )
   } else {
     kv_line("Personal access token", NULL)
+    ui_oops("
+      Call {ui_code('gh_token_help()')} for help configuring a token")
   }
 
   # repo overview -------------------------------------------------------------

--- a/R/git.R
+++ b/R/git.R
@@ -415,6 +415,8 @@ git_remotes <- function() {
 #' git_sitrep()
 #' }
 git_sitrep <- function() {
+  ui_silence(try(proj_get(), silent = TRUE))
+
   # git (global / user) --------------------------------------------------------
   hd_line("Git config (global)")
   kv_line("Name", git_cfg_get("user.name", "global"))
@@ -431,71 +433,11 @@ git_sitrep <- function() {
   hd_line("GitHub")
   default_gh_host <- get_hosturl(default_api_url())
   kv_line("Default GitHub host", default_gh_host)
-  auth_token <- gh::gh_token(api_url = default_gh_host)
-  have_token <- auth_token != ""
-  if (have_token) {
-    kv_line("Personal access token", "<discovered>")
-    tryCatch(
-      {
-        who <- gh::gh_whoami(.token = auth_token, .api_url = default_gh_host)
-        kv_line("GitHub user", who$login)
-        scopes <- who$scopes
-        kv_line("Token scopes", who$scopes)
-        # https://docs.github.com/en/free-pro-team@latest/developers/apps/scopes-for-oauth-apps
-        # why these checks?
-        # previous defaults for create_github_token(): repo, gist, user:email
-        # more recently: repo, user, gist
-        # (gist scope is a very weak recommendation)
-        scopes <- strsplit(scopes, ", ")[[1]]
-        if (length(scopes) == 0 ||
-            !any(grepl("^repo$", scopes)) ||
-            !any(grepl("^user(:email)?$", scopes))) {
-          ui_oops("
-            Token may be mis-scoped: {ui_value('repo')} and \\
-            {ui_value('user')} are highly recommended scopes
-            If you are troubleshooting, consider this")
-        }
-      },
-      http_error_401 = function(e) ui_oops("Token is invalid"),
-      error = function(e) {
-        ui_oops("
-          Can't get user profile for this token. Is the network reachable?")
-      }
-    )
-    tryCatch(
-      {
-        emails <- gh::gh(
-          "/user/emails",
-          .token = auth_token,
-          .api_url = default_gh_host
-        )
-        addresses <- map_chr(
-          emails,
-          ~ if (.x$primary) glue_data(.x, "{email} (primary)") else .x[["email"]]
-        )
-        kv_line("Email(s)", addresses)
-        global_email <- git_cfg_get("user.email", "global")
-        if (!is.null(global_email) && !grepl(global_email, addresses)) {
-          ui_oops(
-            "User's local Git email doesn't appear to be registered with GitHub"
-          )
-        }
-      },
-      error = function(e) {
-        ui_oops("
-          Can't retrieve registered email addresses
-          If you are troubleshooting, check GitHub host, token, and token scopes")
-      }
-    )
-  } else {
-    kv_line("Personal access token", NULL)
-    ui_oops("
-      Call {ui_code('gh_token_help()')} for help configuring a token")
-  }
+  pat_found <- pat_sitrep(host = default_gh_host)
 
   # git and github for active project ------------------------------------------
-  hd_line("Git repo associated with the current project")
-  ui_silence(try(proj_get(), silent = TRUE))
+  hd_line("Git repo for current project")
+
   if (!proj_active()) {
     ui_info("No active usethis project")
     return(invisible())
@@ -513,35 +455,104 @@ git_sitrep <- function() {
       user.email = git_cfg_get("user.email", "local")
     )
     if (!is.null(local_user$user.name) || !is.null(local_user$user.name)) {
-      hd_line("Git config (project)")
+      ui_info("This repo has a locally configured user")
       kv_line("Name", local_user$user.name)
       kv_line("Email", local_user$user.email)
     }
   }
 
+  # default branch -------------------------------------------------------------
+  kv_line("Default branch", git_branch_default())
+
   # current branch -------------------------------------------------------------
   branch <- tryCatch(git_branch(), error = function(e) NULL)
   tracking_branch <- if (is.null(branch)) NA_character_ else git_branch_tracking()
-  # TODO: can't really do what I want with existing ui_*() functions
-  # e.g., vertical alignment would make this nicer
+  # TODO: can't really express with kv_line() helper
   branch <- if (is.null(branch)) "<unset>" else branch
   tracking_branch <- if (is.na(tracking_branch)) "<unset>" else tracking_branch
-  ui_inform(glue("
-    * Local branch -> remote tracking branch:
-      {ui_value(branch)} -> {ui_value(tracking_branch)}"))
+  # vertical alignment would make this nicer, but probably not worth it
+  ui_bullet(glue("
+    Current local branch -> remote tracking branch:
+    {ui_value(branch)} -> {ui_value(tracking_branch)}"), cli::symbol$star)
 
-  # PR outlook -------------------------------------------------------------
-  hd_line("GitHub remote configuration")
-  # TODO: need to surface host more here
+  # GitHub remote config -------------------------------------------------------
   cfg <- github_remote_config()
+  repo_host <- cfg$host_url
+  if (repo_host != default_gh_host) {
+    kv_line("Non-default GitHub host", repo_host)
+    pat_found <- pat_sitrep(host = repo_host)
+  }
+
+  hd_line("GitHub remote configuration")
   if (cfg$type == "no_github") {
     ui_info("
       This repo has neither {ui_value('origin')} nor {ui_value('upstream')} \\
-      remote on GitHub.com.")
+      remote on GitHub")
     return(invisible())
   }
-  # TODO: make this more attractive
   print(cfg)
+}
+
+pat_sitrep <- function(host = "https://github.com") {
+  pat <- gh::gh_token(api_url = host)
+  have_pat <- pat != ""
+  if (!have_pat) {
+    kv_line("Personal access token for {ui_value(host)}", NULL)
+    ui_oops("
+      Call {ui_code('gh_token_help()')} for help configuring a token")
+    return(FALSE)
+  }
+
+  kv_line("Personal access token for {ui_value(host)}", "<discovered>")
+  tryCatch(
+    {
+      who <- gh::gh_whoami(.token = pat, .api_url = host)
+      kv_line("GitHub user", who$login)
+      scopes <- who$scopes
+      kv_line("Token scopes", who$scopes)
+      # https://docs.github.com/en/free-pro-team@latest/developers/apps/scopes-for-oauth-apps
+      # why these checks?
+      # previous defaults for create_github_token(): repo, gist, user:email
+      # more recently: repo, user, gist
+      # (gist scope is a very weak recommendation)
+      scopes <- strsplit(scopes, ", ")[[1]]
+      if (length(scopes) == 0 ||
+          !any(grepl("^repo$", scopes)) ||
+          !any(grepl("^user(:email)?$", scopes))) {
+        ui_oops("
+            Token may be mis-scoped: {ui_value('repo')} and \\
+            {ui_value('user')} are highly recommended scopes
+            If you are troubleshooting, consider this")
+      }
+    },
+    http_error_401 = function(e) ui_oops("Token is invalid"),
+    error = function(e) {
+      ui_oops("
+        Can't get user profile for this token. Is the network reachable?")
+    }
+  )
+  tryCatch(
+    {
+      emails <- gh::gh("/user/emails", .token = pat, .api_url = host)
+      addresses <- map_chr(
+        emails,
+        ~ if (.x$primary) glue_data(.x, "{email} (primary)") else .x[["email"]]
+      )
+      kv_line("Email(s)", addresses)
+      de_facto_email <- git_cfg_get("user.email", "de_facto")
+      if (!any(grepl(de_facto_email, addresses))) {
+        ui_oops("
+          User's Git email ({ui_value(de_facto_email)}) doesn't appear to be \\
+          registered with GitHub")
+      }
+    },
+    error = function(e) {
+      ui_oops("
+        Can't retrieve registered email address(es)
+        If you are troubleshooting, check GitHub host, token, and token scopes")
+    }
+  )
+  TRUE
 }
 
 # Vaccination -------------------------------------------------------------

--- a/R/git.R
+++ b/R/git.R
@@ -415,7 +415,7 @@ git_remotes <- function() {
 #' git_sitrep()
 #' }
 git_sitrep <- function() {
-  # git global ----------------------------------------------------------------
+  # git (global / user) --------------------------------------------------------
   hd_line("Git config (global)")
   kv_line("Name", git_cfg_get("user.name", "global"))
   kv_line("Email", git_cfg_get("user.email", "global"))
@@ -427,20 +427,7 @@ git_sitrep <- function() {
   # TODO: Revisit when I harden the HTTPS default
   kv_line("Default usethis protocol", getOption("usethis.protocol"))
 
-  # git project ---------------------------------------------------------------
-  if (proj_active() && uses_git()) {
-    local_user <- list(
-      user.name = git_cfg_get("user.name", "local"),
-      user.email = git_cfg_get("user.email", "local")
-    )
-    if (!is.null(local_user$user.name) || !is.null(local_user$user.name)) {
-      hd_line("Git config (project)")
-      kv_line("Name", local_user$user.name)
-      kv_line("Email", local_user$user.email)
-    }
-  }
-
-  # github ---------------------------------------------------------------------
+  # github (global / user) -----------------------------------------------------
   hd_line("GitHub")
   default_gh_host <- get_hosturl(default_api_url())
   kv_line("Default GitHub host", default_gh_host)
@@ -451,24 +438,27 @@ git_sitrep <- function() {
     tryCatch(
       {
         who <- gh::gh_whoami(.token = auth_token, .api_url = default_gh_host)
-        kv_line("User", who$login)
-        kv_line("Name", who$name)
+        kv_line("GitHub user", who$login)
         scopes <- who$scopes
-        kv_line("Scopes", who$scopes)
+        kv_line("Token scopes", who$scopes)
+        # why these checks?
         # previous defaults for create_github_token(): repo, gist, user:email
         # more recently: repo, user, gist
-        # and gist scope is a very weak recommendation
-        # hence the checks aren't exactly for our defaults
+        # (gist scope is a very weak recommendation)
         scopes <- strsplit(scopes, ", ")[[1]]
         if (!any(grepl("^repo$", scopes)) ||
-            !any(grepl("user(:email)?", scopes))) {
+            !any(grepl("^user(:email)?$", scopes))) {
           ui_oops("
-            Token may be underscoped: {ui_value('repo')} and \\
-            {ui_value('user')} are highly recommended scopes")
+            Token may be mis-scoped? {ui_value('repo')} and \\
+            {ui_value('user')} are highly recommended scopes
+            If you are troubleshooting, consider this")
         }
       },
       http_error_401 = function(e) ui_oops("Token is invalid."),
-      error = function(e) ui_oops("Can't validate token. Is the network reachable?")
+      error = function(e) {
+        ui_oops("
+          Can't get user profile for this token. Is the network reachable?")
+      }
     )
     tryCatch(
       {
@@ -496,6 +486,19 @@ git_sitrep <- function() {
     kv_line("Personal access token", NULL)
     ui_oops("
       Call {ui_code('gh_token_help()')} for help configuring a token")
+  }
+
+  # git project ---------------------------------------------------------------
+  if (proj_active() && uses_git()) {
+    local_user <- list(
+      user.name = git_cfg_get("user.name", "local"),
+      user.email = git_cfg_get("user.email", "local")
+    )
+    if (!is.null(local_user$user.name) || !is.null(local_user$user.name)) {
+      hd_line("Git config (project)")
+      kv_line("Name", local_user$user.name)
+      kv_line("Email", local_user$user.email)
+    }
   }
 
   # repo overview -------------------------------------------------------------

--- a/R/git.R
+++ b/R/git.R
@@ -419,7 +419,11 @@ git_sitrep <- function() {
   hd_line("Git config (global)")
   kv_line("Name", git_cfg_get("user.name", "global"))
   kv_line("Email", git_cfg_get("user.email", "global"))
-  kv_line("Vaccinated", git_vaccinated())
+  vaccinated <- git_vaccinated()
+  kv_line("Vaccinated", vaccinated)
+  if (!vaccinated) {
+    ui_info("See {ui_code('?git_vaccinate')} to learn more about this")
+  }
 
   # git project ---------------------------------------------------------------
   if (proj_active() && uses_git()) {
@@ -542,9 +546,9 @@ git_sitrep <- function() {
 
 #' Vaccinate your global gitignore file
 #'
-#' Adds `.DS_Store`, `.Rproj.user`, `.Rdata`, and `.Rhistory` to your global
-#' (a.k.a. user-level) `.gitignore`. This is good practice as it decreases the
-#' chance that you will accidentally leak credentials to GitHub.
+#' Adds `.DS_Store`, `.Rproj.user`, `.Rdata`, `.Rhistory`, and `.httr-oauth` to
+#' your global (a.k.a. user-level) `.gitignore`. This is good practice as it
+#' decreases the chance that you will accidentally leak credentials to GitHub.
 #'
 #' @export
 git_vaccinate <- function() {
@@ -566,5 +570,6 @@ git_global_ignore <- c(
   ".Rproj.user",
   ".Rhistory",
   ".Rdata",
+  ".httr-oauth",
   ".DS_Store"
 )

--- a/R/github_token.R
+++ b/R/github_token.R
@@ -55,7 +55,7 @@ NULL
 #' \dontrun{
 #' create_github_token()
 #' }
-create_github_token <- function(scopes = c("repo", "gist", "user:email"),
+create_github_token <- function(scopes = c("repo", "user", "gist"),
                                 description = "R:GITHUB_PAT",
                                 host = NULL) {
   scopes <- glue_collapse(scopes, ",")

--- a/R/ui.R
+++ b/R/ui.R
@@ -289,7 +289,12 @@ ui_unset <- function(x = "unset") {
 
 # rlang::inform() wrappers -----------------------------------------------------
 
-ui_bullet <- function(x, bullet) {
+indent <- function(x, first = "  ", indent = first) {
+  x <- gsub("\n", paste0("\n", indent), x)
+  paste0(first, x)
+}
+
+ui_bullet <- function(x, bullet = cli::symbol$bullet) {
   bullet <- paste0(bullet, " ")
   x <- indent(x, bullet, "  ")
   ui_inform(x)
@@ -311,7 +316,8 @@ hd_line <- function(name) {
   ui_inform(crayon::bold(name))
 }
 
-kv_line <- function(key, value) {
+kv_line <- function(key, value, .envir = parent.frame()) {
   value <- if (is.null(value)) ui_unset() else ui_value(value)
-  ui_inform("* ", key, ": ", value)
+  key <- glue(key, .envir = .envir)
+  ui_inform(glue("{cli::symbol$star} {key}: {value}"))
 }

--- a/R/utils-github.R
+++ b/R/utils-github.R
@@ -1,4 +1,4 @@
-# repo_spec --> owner, repo
+# OWNER/REPO --> OWNER, REPO
 parse_repo_spec <- function(repo_spec) {
   repo_split <- strsplit(repo_spec, "/")[[1]]
   if (length(repo_split) != 2) {
@@ -10,7 +10,7 @@ parse_repo_spec <- function(repo_spec) {
 spec_owner <- function(repo_spec) parse_repo_spec(repo_spec)$owner
 spec_repo <- function(repo_spec) parse_repo_spec(repo_spec)$repo
 
-# owner, repo --> repo_spec
+# OWNER, REPO --> OWNER/REPO
 make_spec <- function(owner = NA, repo = NA) {
   no_spec <- is.na(owner) | is.na(repo)
   as.character(ifelse(no_spec, NA, glue("{owner}/{repo}")))
@@ -167,13 +167,12 @@ github_remotes <- function(these = c("origin", "upstream"),
     oops <- which(!grl$github_got)
     oops_remotes <- grl$remote[oops]
     oops_hosts <- unique(grl$host[oops])
-    # TODO: update when there's a place to send people for troubleshooting
     ui_stop("
       Unable to get GitHub info for these remotes: {ui_value(oops_remotes)}
-      Are we offline?
+      Are we offline? Is GitHub down?
       Otherwise, you probably need to configure a personal access token (PAT) \\
       for {ui_value(oops_hosts)}
-      See {ui_code('?create_github_token')} for advice")
+      See {ui_code('?gh_token_help')} for advice")
   }
 
   grl$is_fork <- map_lgl(repo_info, "fork", .default = NA)
@@ -280,6 +279,7 @@ new_github_remote_config <- function() {
   structure(
     list(
       type = NA_character_,
+      host_url = NA_character_,
       pr_ready = FALSE,
       desc = "Unexpected remote configuration.",
       origin   = c(name = "origin",   is_configured = FALSE, ptype),
@@ -317,6 +317,7 @@ github_remote_config <- function(github_get = NA) {
         Internal error: Know GitHub permissions for some remotes, but not all")
     }
   }
+  cfg$host_url <- unique(grl$host_url)
   github_got <- any(grl$github_got)
   perm_known <- any(grl$perm_known)
 
@@ -511,6 +512,7 @@ format_remote <- function(remote) {
 format_fields <- function(cfg) {
   list(
     type = glue("type = {ui_value(cfg$type)}"),
+    host_url = glue("host = {ui_value(cfg$host_url)}"),
     pr_ready = glue("Config supports a pull request = {ui_value(cfg$pr_ready)}"),
     origin = format_remote(cfg$origin),
     upstream = format_remote(cfg$upstream),

--- a/R/utils-github.R
+++ b/R/utils-github.R
@@ -511,15 +511,15 @@ format_remote <- function(remote) {
 
 format_fields <- function(cfg) {
   list(
-    type = glue("type = {ui_value(cfg$type)}"),
-    host_url = glue("host = {ui_value(cfg$host_url)}"),
+    type = glue("Type = {ui_value(cfg$type)}"),
+    host_url = glue("Host = {ui_value(cfg$host_url)}"),
     pr_ready = glue("Config supports a pull request = {ui_value(cfg$pr_ready)}"),
     origin = format_remote(cfg$origin),
     upstream = format_remote(cfg$upstream),
     desc = if (is.na(cfg$desc)) {
-      glue("desc = {ui_unset('no description')}")
+      glue("Desc = {ui_unset('no description')}")
     } else {
-      glue("desc = {cfg$desc}")
+      glue("Desc = {cfg$desc}")
     }
   )
 }
@@ -618,6 +618,15 @@ check_ours_or_fork <- function(cfg = NULL) {
 }
 
 # github remote configurations -------------------------------------------------
+# use for configs
+read_more <- glue("
+  Read more about the GitHub remote configurations that usethis supports at:
+  {ui_value('https://happygitwithr.com/common-remote-setups.html')}")
+
+read_more_maybe <- glue("
+  Read more about what this GitHub remote configurations means at:
+  {ui_value('https://happygitwithr.com/common-remote-setups.html')}")
+
 cfg_no_github <- function(cfg) {
   utils::modifyList(
     cfg,
@@ -626,7 +635,9 @@ cfg_no_github <- function(cfg) {
       pr_ready = FALSE,
       desc = glue("
         Neither {ui_value('origin')} nor {ui_value('upstream')} is a GitHub \\
-        repo.")
+        repo.
+
+        {read_more}")
     )
   )
 }
@@ -637,7 +648,11 @@ cfg_ours <- function(cfg) {
     list(
       type = "ours",
       pr_ready = TRUE,
-      desc = NA)
+      desc = glue("
+        {ui_value('origin')} is both the source and primary repo.
+
+        {read_more}")
+    )
   )
 }
 
@@ -652,7 +667,9 @@ cfg_theirs <- function(cfg) {
         The only configured GitHub remote is {ui_value(configured)}, which
         you cannot push to.
         If your goal is to make a pull request, you must fork-and-clone.
-        {ui_code('usethis::create_from_github()')} can do this.")
+        {ui_code('usethis::create_from_github()')} can do this.
+
+        {read_more}")
     )
   )
 }
@@ -672,7 +689,12 @@ cfg_maybe_ours_or_theirs <- function(cfg) {
       pr_ready = NA,
       desc = glue("
         {ui_value(configured)} is a GitHub repo and {ui_value(not_configured)} \\
-        is either not configured or is not a GitHub repo.")
+        is either not configured or is not a GitHub repo.
+
+        We may be offline or you may need to configure a GitHub personal access
+        token. {ui_code('gh_token_help()')} can help with that.
+
+        {read_more_maybe}")
     )
   )
 }
@@ -683,10 +705,11 @@ cfg_fork <- function(cfg) {
     list(
       type = "fork",
       pr_ready = TRUE,
-      # TODO: say whether user can push to parent / upstream?
       desc = glue("
         {ui_value('origin')} is a fork of {ui_value(cfg$upstream$repo_spec)}, \\
-        which is configured as the {ui_value('upstream')} remote.")
+        which is configured as the {ui_value('upstream')} remote.
+
+        {read_more}")
     )
   )
 }
@@ -701,8 +724,12 @@ cfg_maybe_fork <- function(cfg) {
         Both {ui_value('origin')} and {ui_value('upstream')} appear to be \\
         GitHub repos. However, we can't confirm their relationship to each \\
         other (e.g., fork and fork parent) or your permissions (e.g. push \\
-        access). We may be offline or you may need to configure a GitHub \\
-        personal access token.")
+        access).
+
+        We may be offline or you may need to configure a GitHub personal access
+        token. {ui_code('gh_token_help()')} can help with that.
+
+        {read_more_maybe}")
     )
   )
 }
@@ -714,7 +741,9 @@ cfg_fork_cannot_push_origin <- function(cfg) {
       type = "fork_cannot_push_origin",
       pr_ready = FALSE,
       desc = glue("
-        The {ui_value('origin')} remote is a fork, but you can't push to it.")
+        The {ui_value('origin')} remote is a fork, but you can't push to it.
+
+        {read_more}")
     )
   )
 }
@@ -727,7 +756,9 @@ cfg_fork_upstream_is_not_origin_parent <- function(cfg) {
       pr_ready = FALSE,
       desc = glue("
         The {ui_value('origin')} GitHub remote is a fork, but its parent is \\
-        not configured as the {ui_value('upstream')} remote.")
+        not configured as the {ui_value('upstream')} remote.
+
+        {read_more}")
     )
   )
 }
@@ -741,7 +772,9 @@ cfg_upstream_but_origin_is_not_fork <- function(cfg) {
       desc = glue("
         Both {ui_value('origin')} and {ui_value('upstream')} are GitHub \\
         remotes, but {ui_value('origin')} is not a fork and, in particular, \\
-        is not a fork of {ui_value('upstream')}.")
+        is not a fork of {ui_value('upstream')}.
+
+        {read_more}")
     )
   )
 }

--- a/R/utils.R
+++ b/R/utils.R
@@ -61,11 +61,6 @@ interactive <- function() {
   )
 }
 
-indent <- function(x, first = "  ", indent = first) {
-  x <- gsub("\n", paste0("\n", indent), x)
-  paste0(first, x)
-}
-
 isFALSE = function(x) {
   identical(x, FALSE)
 }

--- a/man/git_vaccinate.Rd
+++ b/man/git_vaccinate.Rd
@@ -7,7 +7,7 @@
 git_vaccinate()
 }
 \description{
-Adds \code{.DS_Store}, \code{.Rproj.user}, \code{.Rdata}, and \code{.Rhistory} to your global
-(a.k.a. user-level) \code{.gitignore}. This is good practice as it decreases the
-chance that you will accidentally leak credentials to GitHub.
+Adds \code{.DS_Store}, \code{.Rproj.user}, \code{.Rdata}, \code{.Rhistory}, and \code{.httr-oauth} to
+your global (a.k.a. user-level) \code{.gitignore}. This is good practice as it
+decreases the chance that you will accidentally leak credentials to GitHub.
 }

--- a/man/github-token.Rd
+++ b/man/github-token.Rd
@@ -7,7 +7,7 @@
 \title{Get help with GitHub personal access tokens}
 \usage{
 create_github_token(
-  scopes = c("repo", "gist", "user:email"),
+  scopes = c("repo", "user", "gist"),
   description = "R:GITHUB_PAT",
   host = NULL
 )

